src/util/secrets/secrets.c

```diff
@@ -0,0 +1,1235 @@
+              ret, sss_strerror(ret));
+        goto done;
+    }
+
+    ret = ldb_msg_add_string(msg, "secret", enc_secret);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "ldb_msg_add_string failed adding secret [%d]: %s\n",
+              ret, sss_strerror(ret));
+        goto done;
+    }
+
+
+    ret = ldb_msg_add_fmt(msg, "creationTime", "%lu", time(NULL));
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "ldb_msg_add_string failed adding creationTime [%d]: %s\n",
+              ret, sss_strerror(ret));
+        goto done;
+    }
+
+    ret = ldb_add(req->sctx->ldb, msg);
+    if (ret != EOK) {
+        if (ret == LDB_ERR_ENTRY_ALREADY_EXISTS) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "Secret %s already exists\n", ldb_dn_get_linearized(msg->dn));
+            ret = EEXIST;
+        } else {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "Failed to add secret [%s]: [%d]: %s\n",
+                  ldb_dn_get_linearized(msg->dn), ret, ldb_strerror(ret));
+            ret = EIO;
+        }
+        goto done;
+    }
+
+    ret = EOK;
+done:
+    talloc_free(msg);
+    return ret;
+}
+
+errno_t sss_sec_delete(struct sss_sec_req *req)
+{
+    TALLOC_CTX *tmp_ctx;
+    static const char *attrs[] = { NULL };
+    struct ldb_result *res;
+    int ret;
+
+    if (req == NULL) {
+        return EINVAL;
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Removing a secret from [%s]\n", req->path);
+
+    tmp_ctx = talloc_new(req);
+    if (!tmp_ctx) return ENOMEM;
+
+    DEBUG(SSSDBG_TRACE_INTERNAL,
+          "Searching for [%s] at [%s] with scope=base\n",
+          LOCAL_CONTAINER_FILTER, ldb_dn_get_linearized(req->req_dn));
+
+    ret = ldb_search(req->sctx->ldb, tmp_ctx, &res, req->req_dn, LDB_SCOPE_BASE,
+                     attrs, LOCAL_CONTAINER_FILTER);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_TRACE_LIBS,
+              "ldb_search returned %d: %s\n", ret, ldb_strerror(ret));
+        goto done;
+    }
+
+    if (res->count == 1) {
+        DEBUG(SSSDBG_TRACE_INTERNAL,
+              "Searching for children of [%s]\n", ldb_dn_get_linearized(req->req_dn));
+        ret = ldb_search(req->sctx->ldb, tmp_ctx, &res, req->req_dn, LDB_SCOPE_ONELEVEL,
+                         attrs, NULL);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_TRACE_LIBS,
+                  "ldb_search returned %d: %s\n", ret, ldb_strerror(ret));
+            goto done;
+        }
+
+        if (res->count > 0) {
+            ret = EEXIST;
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "Failed to remove '%s': Container is not empty\n",
+                  ldb_dn_get_linearized(req->req_dn));
+
+            goto done;
+        }
+    }
+
+    ret = ldb_delete(req->sctx->ldb, req->req_dn);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_TRACE_LIBS,
+              "ldb_delete returned %d: %s\n", ret, ldb_strerror(ret));
+        /* fall through */
+    }
+
+    switch (ret) {
```

Comment:
Don't we already have a utility function to convert ldb errors to sss errors?

Reply:
Yes, but currently it's in sysdb and has over 200 callers. I think it might be OK to move the function to utiil/debug.c and call it perhaps `ldb_err_to_sss_err` but I don't like the churn the rename would take. Especially if we push these patches to 2.x only, then backports would get more difficult.

Or, we could rename and move the function and then `#define sysdb_error_to_errno ldb_err_to_sss_err`.

Reply:
I see. Okay, leave it as it is for now.

Reply:
https://pagure.io/SSSD/sssd/issue/3802 - might be a good task for newcomers to the team